### PR TITLE
Initialize

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,136 @@
...
+    21.0
+
+A value can also be retrieved as textual output via string formatting. The
+syntax is similar to that of floating point values, with the desired unit
+taking the place of the 'f' suffix.
+
+    >>> f'velocity: {v:.1m/s}'
+    'velocity: 21.0m/s'
+
+A Quantity container can hold an object of any type that supports arithmatical
```
Comment (joostvanzwieten) [inline README.md:74]: arithmetical

File: README.md [hunk 2]
```diff
@@ -1,2 +1,136 @@
...
+At this point, the dimension is not very useful yet as it lacks units. To
+rectify this we define the radian by its abbreviation 'rad' and assign it to
+the global table of units:
+
+    >>> SI.units.rad = Angle(1.)
+
+As Quantity objects are opaque, the number 1 here is entirely arbitrary, since
+any choice will ultimately be cancelled out when the value is retrieved by
+dividing out a unit or other reference value. It does, however, have internal
+significance, is it lends meaning to the value that is internal to an Angle
```
Comment (joostvanzwieten) [inline README.md:111]: is => as (it lends ...)

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    strategy:
+      matrix:
+        include:
+          - {os: ubuntu-latest,  python-version: 3.5}
+          - {os: ubuntu-latest,  python-version: 3.6}
+          - {os: ubuntu-latest,  python-version: 3.7}
+          - {os: ubuntu-latest,  python-version: 3.8}
+          - {os: ubuntu-latest,  python-version: 3.9}
+          - {os: windows-latest, python-version: 3.9}
+          - {os: macos-latest,   python-version: 3.9}
```
Comment (joostvanzwieten) [inline .github/workflows/test.yaml:23]: Should we add 3.10 as well?

File: nutils/SI.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py]: done

File: nutils/SI.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+import operator, typing, fractions
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py]: done

File: nutils/SI.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+
+class Dimension(type):
+
+    __cache = {} # subtypes
+
+    def __new__(mcls, *args):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py]: done via introduction of create and from_powers

File: nutils/SI.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        if len(args) == 3:
+            return type.__new__(mcls, *args)
+        # manual instantiation of Dimension subtype
+        arg, = args
+        if isinstance(arg, str):
+            powers = {base: fractions.Fraction(power if isnumer else -power) for base, power, isnumer in _split_factors(arg) if power}
+        elif isinstance(arg, dict):
+            powers = {base: fractions.Fraction(power) for base, power in arg.items() if power}
+        else:
+            raise ValueError('cannot create Dimension from {!r}'.format(name))
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py]: done

File: nutils/SI.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+        name = ''.join(('*' if power > 0 else '/') + base
+                     + (str(abs(power.numerator)) if abs(power.numerator) != 1 else '')
+                     + ('_'+str(abs(power.denominator)) if abs(power.denominator) != 1 else '')
+            for base, power in sorted(powers.items(), key=lambda item: item[::-1], reverse=True)).lstrip('*')
+        try:
+            cls = mcls.__cache[name]
+        except KeyError:
+            cls = type.__new__(mcls, '[{}]'.format(name), (Quantity,), {})
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py]: done

File: nutils/SI.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+        return v.__format__(format_spec[:n]+'f') + format_spec[n:]
+
+    @staticmethod
+    def _dispatch(op, *args, **kwargs):
+        name = op.__name__
+        if name in ('add', 'iadd', 'sub', 'isub', 'subtract', 'hypot'):
+            Dim = type(args[0])
+            if type(args[1]) != Dim:
+                raise TypeError('incompatible arguments for {}: {}'.format(name, ', '.join(type(arg).__name__ for arg in args)))
+        elif name in ('mul', 'imul', 'multiply', 'matmul', 'imatmul'):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/SI.py:140]: removed all in place operations

File: README.md (hunk 1 above)
Comment (gertjanvanzwieten) [reply README.md:74]: fixed

File: README.md (hunk 2 above)
Comment (gertjanvanzwieten) [reply README.md:111]: fixed

File: .github/workflows/test.yaml (hunk 3 above)
Comment (gertjanvanzwieten) [reply .github/workflows/test.yaml:23]: done

**Comment by gertjanvanzwieten** [comment]
Remaining to-do items:
- add unit tests for str operations (q / 'mm') plus readme entry
- decide on name for SI.units / SI.unit
- maybe be more strict in Quantity._dispatch

**Pull request merged**